pkg/querier/querier.go

```diff
@@ -102,52 +91,56 @@ type Querier interface {
 	logql.Querier
 	Label(ctx context.Context, req *logproto.LabelRequest) (*logproto.LabelResponse, error)
 	Series(ctx context.Context, req *logproto.SeriesRequest) (*logproto.SeriesResponse, error)
-	Tail(ctx context.Context, req *logproto.TailRequest, categorizedLabels bool) (*Tailer, error)
 	IndexStats(ctx context.Context, req *loghttp.RangeQuery) (*stats.Stats, error)
 	IndexShards(ctx context.Context, req *loghttp.RangeQuery, targetBytesPerShard uint64) (*logproto.ShardsResponse, error)
 	Volume(ctx context.Context, req *logproto.VolumeRequest) (*logproto.VolumeResponse, error)
 	DetectedFields(ctx context.Context, req *logproto.DetectedFieldsRequest) (*logproto.DetectedFieldsResponse, error)
 	Patterns(ctx context.Context, req *logproto.QueryPatternsRequest) (*logproto.QueryPatternsResponse, error)
 	DetectedLabels(ctx context.Context, req *logproto.DetectedLabelsRequest) (*logproto.DetectedLabelsResponse, error)
-	WithPatternQuerier(patternQuerier PatterQuerier)
+	WithPatternQuerier(patternQuerier pattern.PatterQuerier)
 }
 
-type Limits querier_limits.Limits
-
 // Store is the store interface we need on the querier.
 type Store interface {
-	storage.SelectStore
-	index.BaseReader
-	index.StatsReader
+	SelectSamples(ctx context.Context, req logql.SelectSampleParams) (iter.SampleIterator, error)
+	SelectLogs(ctx context.Context, req logql.SelectLogParams) (iter.EntryIterator, error)
+	SelectSeries(ctx context.Context, req logql.SelectLogParams) ([]logproto.SeriesIdentifier, error)
+	LabelValuesForMetricName(ctx context.Context, userID string, from, through model.Time, metricName string, labelName string, matchers ...*labels.Matcher) ([]string, error)
+	LabelNamesForMetricName(ctx context.Context, userID string, from, through model.Time, metricName string, matchers ...*labels.Matcher) ([]string, error)
+	Stats(ctx context.Context, userID string, from, through model.Time, matchers ...*labels.Matcher) (*stats.Stats, error)
+	Volume(ctx context.Context, userID string, from, through model.Time, limit int32, targetLabels []string, aggregateBy string, matchers ...*labels.Matcher) (*logproto.VolumeResponse, error)
+	GetShards(
+		ctx context.Context,
+		userID string,
+		from, through model.Time,
+		targetBytesPerShard uint64,
+		predicate chunk.Predicate,
+	) (*logproto.ShardsResponse, error)
```

Comment:
I often struggle to understand when interface composition and when separate definition is better. From the PR alone it's not clear to me why making the function definitions separately in this interface has advantages. Could you elaborate? I assume this is kinda necessary for future work?

Reply:
This change reduces the Store interface to only include methods actually used by the querier.
Previously, the interface included unused methods from storage.SelectStore and
index interfaces, creating unnecessary coupling between packages.

Following Go's interface segregation principle, interfaces should only declare
methods that clients need. This makes the code more maintainable since:
1. Changes to unused methods in storage/index packages won't affect the querier
2. The querier's dependencies are explicitly visible through its minimal interface

Interface composition is still valuable when building upon well-known Go
or project wide patterns (like io.Writer or iter.Iterator), but wasn't necessary in this case.

Reply:
Thanks!